### DISCUSSION TITLE
Provide a example about how to use ArrayState

### DISCUSSION THREAD

**Discussion by MasonAlice** [vsg-dev/VulkanSceneGraph#701] [Q&A] [answered]
I found `StateGroup` use` ArrayState` to compute bounding box or to intersect. But I don't know how to use it clearly. is there any example about it.

**Comment by robertosfield** [comment, marked as answer]
ArrayState is essentially helper class for handling scene graphs that use vertex shaders that process vertices in non standard ways i,e move the vertices around in ways that are as simple as modelmatrix * vertex. They provide a form CPU emulation of vertex shader so CPU routines can process vertex and primitive data without knowing the details of every possible shader combination.


There isn't an vsgExamples example of using vsg::ArrayState, but both the [vsg::ComputeBounds](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/utils/ComputeBounds.cpp)  and v[sg::LineSegmentIntersector](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/utils/LineSegmentIntersector.cpp) implementations found in the VulkanSceneGraph use it so may be helpful.

There are also implementations of custom [ArrayState](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/state/ArrayState.cpp#L184) provided by the built in ShaderSets to handle shaders that move the vertices around such as displacement maps etc.

What task are you looking to tackle that you think ArrayState will be helpful?

**Comment by vsg-dev** [comment]
@HowellDong I am moving this question to the Disucssion pages as this is the appropriate place to ask questions.  The Issue tracker is for build and runtime issues not general questions how to do things.

**Comment by MasonAlice** [comment]
> ArrayState is essentially helper class for handling scene graphs that use vertex shaders that process vertices in non standard ways i,e move the vertices around in ways that are as simple as modelmatrix * vertex. They provide a form CPU emulation of vertex shader so CPU routines can process vertex and primitive data without knowing the details of every possible shader combination.
> 
> There isn't an vsgExamples example of using vsg::ArrayState, but both the [vsg::ComputeBounds](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/utils/ComputeBounds.cpp) and v[sg::LineSegmentIntersector](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/utils/LineSegmentIntersector.cpp) implementations found in the VulkanSceneGraph use it so may be helpful.
> 
> There are also implementations of custom [ArrayState](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/state/ArrayState.cpp#L184) provided by the built in ShaderSets to handle shaders that move the vertices around such as displacement maps etc.
> 
> What task are you looking to tackle that you think ArrayState will be helpful?

 I am trying to parse gltf model with tiny-gltf library. I am confused that how vsg::LineSegmentIntersector or ComputeBound know the infomation about vertex's position attribute. (I know vsgXchange using assimp to parse model files, but I want to do a excise. It make me understand vsg better). 
or briefly. how can I make a drawable to intersect with line segment correctly.

**Reply by robertosfield** [reply]
That's an interesting task, any chance you'll share your experiments with the community?  This could help others that are also learning, and also make it possible for others to review, make suggests and contributions to it.

If you are using vsg::ShaderSet/GraphicsPipelineConfigurator with the built in ShaderSet like the vsgXchange::assimp loader does then you shouldn't need to worry about vsg::ArrayState as the appropriate one will be assigned for you.
